File: lib/features/chat/models/message.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'message.freezed.dart';
part 'message.g.dart';

@freezed
class Message with _$Message {
  const factory Message({
    required int id,
    required int conversationId,
    required String senderId,
    String? content,
    String? mediaUrl,
    String? mediaType,
    @Default('text') String type,
    required DateTime createdAt,
    DateTime? editedAt,
    @Default(false) bool isDeleted,
    String? senderName,
    String? senderAvatar,
    @Default(false) bool isRead,
    @Default([]) List<MessageReaction> reactions,
  }) = _Message;

  factory Message.fromJson(Map<String, dynamic> json) =>
      _$MessageFromJson(json);
}

@freezed
class CreateMessage with _$CreateMessage {
  const factory CreateMessage({
    String? content,
    String? mediaUrl,
    String? mediaType,
    @Default('text') String type,
  }) = _CreateMessage;

  factory CreateMessage.fromJson(Map<String, dynamic> json) =>
      _$CreateMessageFromJson(json);
}

@freezed
class MessageReaction with _$MessageReaction {
  const factory MessageReaction({
    required String reaction,
    required String userId,
    String? userName,
    required DateTime createdAt,
  }) = _MessageReaction;

  factory MessageReaction.fromJson(Map<String, dynamic> json) =>
      _$MessageReactionFromJson(json);
}

File: MoviePlusBackend/MoviePlusApi/Models/Chat/Message.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace MoviePlusApi.Models.Chat
{
    public class Message
    {
        [Key]
        public long Id { get; set; }
        
        [Required]
        public int ConversationId { get; set; }
        
        [Required]
        public Guid SenderId { get; set; }
        
        public string? Content { get; set; }
        
        [MaxLength(1000)]
        public string? MediaUrl { get; set; }
        
        [MaxLength(50)]
        public string? MediaType { get; set; }
        
        [Required, MaxLength(50)]
        public string Type { get; set; } = "text";
        
        [Required]
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        
        public DateTime? EditedAt { get; set; }
        
        [Required]
        public bool IsDeleted { get; set; } = false;
        
        // Navigation properties
        [ForeignKey("ConversationId")]
        public virtual Conversation Conversation { get; set; } = null!;
        
        [ForeignKey("SenderId")]
        public virtual User? Sender { get; set; }
        
        public virtual ICollection<MessageReadReceipt> ReadReceipts { get; set; } = new List<MessageReadReceipt>();
        public virtual ICollection<MessageReaction> Reactions { get; set; } = new List<MessageReaction>();
    }
}
